### PR TITLE
Implement choice type

### LINKED ISSUE #114: Add choice type
The choice type determines if a string is in a list of possible choices (name from to the python optparse choice option). This would replace the current heavy usage of `string with match(SELF, '^(a|b|c)$')` with

``` pan
    "attr" ? choice('a', 'b', 'c')
```

String as a basetype is the most common one (or even only one). makes no sense for boolean i think, and long already has `long(1..4)` which is something similar. so only float is not covered, which i think is acceptable.

Related to https://github.com/quattor/template-library-core/pull/102

To be implemented via one of
- make a new type buitin that does this (we can use #95 in case someone wants the `choice` to be usabel via a function `function is_choice = {is_type(choice, ARGV);}` 
- a new builtin `is_choice` and we can make a pantype `type choice = string[] with is_choice(SELF)`

### PR DESCRIPTION
**Pull request by iliclaey**
Fixes #114 

> The choice type determines if a string is in a list of possible choices (name from to the python optparse choice option). This would replace the current heavy usage of `string with match(SELF, '^(a|b|c)$')` with
> 
> ``` pan
>     "attr" ? choice('a', 'b', 'c')
> ```
> 
> String as a basetype is the most common one (or even only one). makes no sense for boolean i think, and long already has `long(1..4)` which is something similar. so only float is not covered, which i think is acceptable.

### CODE REVIEW

File: panc-docs/source/pan-book/pan-book.rst [hunk 1]
```diff
@@ -2140,6 +2140,18 @@ children can be a full type specification and may contain default values
 and/or validation blocks. One can also attach default values or
 validation blocks to the record as a whole.
 
+The choice-type
+~~~~~~~~~~~~~~~
+The choice type is an advanced type that can be used to create user-defined types.
+It takes two or more strings as input arguments. These arguments represent the
```
Comment (stdweird) [inline panc-docs/source/pan-book/pan-book.rst]: it's true that choice with one argument makes not much sense, but it should be possible (i assume it is). re rephrase this to `one or more`

File: panc/src/main/java/org/quattor/pan/dml/functions/Value.java [hunk 2]
```diff
@@ -95,4 +95,7 @@ public Element execute(Context context) {
 		}
 	}
 
+	public String toString() {
```
Comment (stdweird) [inline panc/src/main/java/org/quattor/pan/dml/functions/Value.java]: why is this needed?

File: panc/src/main/java/org/quattor/pan/type/ChoiceType.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    public ChoiceType(String source, SourceRange sourceRange, List<Element> choices) {
+        super(source, sourceRange, "string");
+        this.choices = choices;
+    }
+
+    public static ChoiceType getInstance(String source, SourceRange sourceRange, List<Element> choices)
+            throws SyntaxException {
+
+        // There need to be at least two possible options to choose from.
```
Comment (stdweird) [inline panc/src/main/java/org/quattor/pan/type/ChoiceType.java]: ok, i would not restrict this (altough it makes not much sense to use it with one argument).

File: panc/src/main/java/org/quattor/pan/type/ChoiceType.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    public Object validate(final Context context, final Element self) {
+
+        // Check if the value is indeed a StringProperty.
+        FullType type = context.getFullType(identifier);
+        type.validate(context, self);
+        String source = ((StringProperty) self).getValue();
+
+        // Check whether the value is one of the possible choices.
+        boolean found = false;
+        for (Element e : choices) {
```
Comment (stdweird) [inline panc/src/main/java/org/quattor/pan/type/ChoiceType.java]: java has no builtin method to do this? (ho wi sth epan function `index` implemented?)

File: panc/src/main/java/org/quattor/pan/type/ChoiceType.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("choice(");
+
+        String separator = "";
+        for (Element e : choices) {
+            sb.append(separator);
+            sb.append(((StringProperty) e).getValue());
+            separator = ", ";
```
Comment (stdweird) [inline panc/src/main/java/org/quattor/pan/type/ChoiceType.java]: like `join`?

File: panc/src/main/jjtree/PanParser.jjt [hunk 6]
```diff
@@ -706,6 +708,13 @@ org.quattor.pan.template.SourceRange typeSpec() #TypeSpec :
 org.quattor.pan.template.SourceRange baseTypeSpec() #BaseTypeSpec :
 { Token t = null, b=null, e=null; org.quattor.pan.utils.Range r = null; }
 {
+  t=<CHOICE> <LPAREN> stringLiteral() (<COMMA> stringLiteral())* e=<RPAREN>
+  {
+    jjtThis.setIdentifier(t.image);
+    jjtThis.setSourceRange(PanParserUtils.sourceRangeFromTokens(t,(e!=null)?e:t));
```
Comment (stdweird) [inline panc/src/main/jjtree/PanParser.jjt]: spaces after `,`, and round `?` adn `:`

File: panc/src/main/resources/org/quattor/pan/Messages.properties [hunk 7]
```diff
@@ -231,3 +231,4 @@ MSG_INVALID_RADIX=radix value {0} is not within valid range [{1}, {2}]
 MSG_INVALID_RADIX_TYPE=radix must be a long not {0}
 MSG_INVALID_STRING_FOR_LONG=string cannot be converted to long: {0}
 MSG_OPERATION_WITHOUT_CONTEXT=operation called without context object
+MSG_INVALID_CHOICE_TYPE=value ({0}) is not a possible choice value
```
Comment (stdweird) [inline panc/src/main/resources/org/quattor/pan/Messages.properties]: add newline at EOF

File: panc/src/test/pan/Functionality/choice/choice1.pan [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#
+# @expect="/nlist[@name='profile']/string[@name='x']='a'"
+#
+
+object template choice1;
+
+type mychoice = choice("a", "b", "c");
+
+bind '/x' = mychoice;
+'/x' = "a";
```
Comment (stdweird) [inline panc/src/test/pan/Functionality/choice/choice1.pan]: newline at EOF

File: panc/src/test/pan/Functionality/choice/choice10.pan [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# Test HashType with ChoiceType
+# @expect="/nlist[@name='profile']/nlist[@name='x']/string[@name='entry1']='b' and /nlist[@name='profile']/nlist[@name='x']/string[@name='entry2']='c'"
+#
+
+object template choice10;
+
+type mychoice = choice("a", "b", "c"){};
+bind '/x' = mychoice;
+'/x' = dict("entry1", "b", "entry2", "c");
```
Comment (stdweird) [inline panc/src/test/pan/Functionality/choice/choice10.pan]: newline at all EOF s below as well

File: panc/src/test/pan/Functionality/choice/choice18.pan [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#
+# @expect="/nlist[@name='profile']/string[@name='x']='aaa'"
+#
+
+object template choice18;
+
+type mychoice = choice("aa", "bb", "aaa", "bbb");
+bind '/x' = mychoice with length(SELF) == 3;
```
Comment (stdweird) [inline panc/src/test/pan/Functionality/choice/choice18.pan]: can you also add a test with the `mychoice(3)`?

File: panc/src/test/pan/Functionality/choice/choice20.pan [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#
+# @expect=org.quattor.pan.exceptions.ValidationException
```
Comment (stdweird) [inline panc/src/test/pan/Functionality/choice/choice20.pan]: is there a way to distinguish the exception between this failure and the on ein choice19?

**Comment by jrha** [comment]
This title of this PR should be imperative, i.e. "Implement choice type", copying some of the description from the linked issue would also be helpful to people reading change logs later on.
